### PR TITLE
l2geth: forward tx to sequencer

### LINKED ISSUE #1781: Have replicas forward write requests to the Sequencer
**Is your feature request related to a problem? Please describe.**
Infrastructure providers (Infura, Alchemy, QuickNode, etc.) currently forward write requests to the Sequencer with custom proxy infrastructure that inspects the JSON-RPC request and checks the method. I have received various requests to build this functionality into the replica node software so that other people can run replicas too and can choose not to rely on infra providers for write requests.

**Describe the solution you'd like**
Modify L2Geth so that replica nodes will forward write requests to the Sequencer if configured with a Sequencer RPC url.

### PR DESCRIPTION
**Pull request by tuxcanfly**
**Description**

This PR makes it easier for replica nodes to handle a transaction submitted via RPC and to transparently forward it to the sequencer, based on the configured endpoint.

This allows the consumer of the replica node's RPC to directly submit transactions instead of having to rely on third-party applications.

**Metadata**
- Fixes #1781

### CODE REVIEW

File: l2geth/internal/ethapi/api.go [hunk 1]
```diff
@@ -1647,18 +1648,30 @@ func (s *PublicTransactionPoolAPI) FillTransaction(ctx context.Context, args Sen
...
 	}
 
-	tx := new(types.Transaction)
-	if err := rlp.DecodeBytes(encodedTx, tx); err != nil {
-		return common.Hash{}, err
+	// Forward tx to sequencer if config is set
+	// TODO: add forwarding config
+	if s.b.IsVerifier() {
+		// TODO: replace with env var L2_URL
+		client, err := ethclient.Dial("http://l2geth:8545")
```
Comment (tuxcanfly) [inline l2geth/internal/ethapi/api.go:1664]: Will add timeout and/or verify chainid before booting up the replica at all.

File: l2geth/internal/ethapi/api.go [hunk 2]
```diff
@@ -1647,18 +1648,30 @@ func (s *PublicTransactionPoolAPI) FillTransaction(ctx context.Context, args Sen
...
 		return common.Hash{}, errors.New("Cannot send raw transaction while syncing")
 	}
 
-	tx := new(types.Transaction)
-	if err := rlp.DecodeBytes(encodedTx, tx); err != nil {
-		return common.Hash{}, err
+	// Forward tx to sequencer if config is set
+	// TODO: add forwarding config
+	if s.b.IsVerifier() {
+		// TODO: replace with env var L2_URL
```
Comment (tynes) [inline l2geth/internal/ethapi/api.go:1663]: An example for how to parse and pass through the config is here:
- https://github.com/ethereum-optimism/optimism/blob/6749ab3d3c8d07c9d4ca4a8b4eac06a9380eefc4/l2geth/cmd/utils/flags.go#L814
- https://github.com/ethereum-optimism/optimism/blob/6749ab3d3c8d07c9d4ca4a8b4eac06a9380eefc4/l2geth/cmd/utils/flags.go#L1104
- https://github.com/ethereum-optimism/optimism/blob/6749ab3d3c8d07c9d4ca4a8b4eac06a9380eefc4/l2geth/eth/backend.go#L227

Comment (tuxcanfly) [reply l2geth/internal/ethapi/api.go]: Thanks, using that now.

File: l2geth/internal/ethapi/api.go (hunk 1 above)
Comment (tuxcanfly) [reply l2geth/internal/ethapi/api.go]: Added timeout.

**Review by mslipper — APPROVED** [review]

**Comment by tynes** [comment]
This looks good to me, some thoughts:
- We should have an integration test for this functionality, its really easy to add. This functionality should be preserved in the 1.0 release due to the public mempool being added back to the system. The logic would be similar to this: https://github.com/ethereum-optimism/optimism/blob/f3850f3103d1e35efe185a9e5a85bae9a3b23b5d/integration-tests/test/replica.spec.ts#L54

```js
 const signed = await env.l2Wallet.signTransaction(tx)
 const result = await env.replicaProvider.sendTransaction(signed)

let receipt: TransactionReceipt
while (!receipt) {
  receipt = await env.replicaProvider.getTransactionReceipt(result.hash)
  await sleep(200)
}
```

- Once https://github.com/ethereum-optimism/optimism/pull/2007, then the RPC methods that manage user keys should probably also be updated to proxy (`eth_sendTransaction`). I don't think that this is a super high priority as its unlikely many people will use it, but it would better match people's expectations on how upstream works.

**Review by tynes — CHANGES_REQUESTED** [review]
Good work! Will approve once the integration test is added :)

**Review by tynes — APPROVED** [review]

**Pull request merged**